DifferentiationInterface/docs/src/overloads.md

```diff
@@ -0,0 +1,93 @@
+# Table of overloads
+
+This table recaps the features of each extension, with respect to high-level operators.
+Each cell can have two values
+
+- ❌: the backend does not support this operator
+- ✅: our extension calls the backend operator and handles preparation if possible
+
+Checkmarks (✅) are clickable and link to the source code.
+
+```@setup overloads
+using ADTypes
+using DifferentiationInterface
+using DifferentiationInterface: backend_string
+import Markdown
+import Diffractor, Enzyme, FastDifferentiation, FiniteDiff, FiniteDifferences, ForwardDiff, PolyesterForwardDiff, ReverseDiff, Tapir, Tracker, Zygote
+
+ext_module(ext::Symbol) = Base.get_extension(DifferentiationInterface, ext)
+
+function all_backends_and_extensions()
+    return [
+        (AutoDiffractor(), ext_module(:DifferentiationInterfaceDiffractorExt)),
+        (AutoEnzyme(; mode=Enzyme.Forward), ext_module(:DifferentiationInterfaceEnzymeExt)),
+        (AutoEnzyme(; mode=Enzyme.Reverse), ext_module(:DifferentiationInterfaceEnzymeExt)),
+        (AutoFastDifferentiation(), ext_module(:DifferentiationInterfaceFastDifferentiationExt)),
+        (AutoFiniteDiff(), ext_module(:DifferentiationInterfaceFiniteDiffExt)),
+        (AutoFiniteDifferences(; fdm=FiniteDifferences.central_fdm(3, 1)), ext_module(:DifferentiationInterfaceFiniteDifferencesExt)),
+        (AutoForwardDiff(), ext_module(:DifferentiationInterfaceForwardDiffExt)),
+        (AutoPolyesterForwardDiff(; chunksize=1), ext_module(:DifferentiationInterfacePolyesterForwardDiffExt)),
+        (AutoReverseDiff(), ext_module(:DifferentiationInterfaceReverseDiffExt)),
+        (AutoTapir(), ext_module(:DifferentiationInterfaceTapirExt)),
+        (AutoTracker(), ext_module(:DifferentiationInterfaceTrackerExt)),
+        (AutoZygote(), ext_module(:DifferentiationInterfaceZygoteExt)),
+    ]
+end
+
+operators = (
+    :value_and_derivative!, 
+    :value_and_derivative, 
+    :derivative!, 
+    :derivative,   
+    :value_and_gradient!, 
+    :value_and_gradient, 
+    :gradient!, 
+    :gradient, 
+    :value_and_jacobian!, 
+    :value_and_jacobian, 
+    :jacobian!, 
+    :jacobian, 
+    :hvp!, 
+    :hvp, 
+    :hessian!,
+    :hessian, 
+)
+
+function method_overloaded(operator::Symbol, argtypes, m::Module)
+    f = @eval DifferentiationInterface.$operator
+    ms = methods(f, argtypes, m)
+
+    n = length(ms)
+    n == 0 && return "❌"
+    n == 1 && return "[✅]($(Base.url(only(ms))))"
+    return "✅"
+end
+
+io = IOBuffer()
+
+for (backend, ext) in all_backends_and_extensions()
+    bname = backend_string(backend)
+    btype = typeof(backend)
+
+    # Subsection title
+    println(io, "## $bname")
+    # Table header
+    println(io, "| Operator | `operator(f, backend, x, ...)`| `operator(f!, y, backend, x, ...)` |")
+    println(io, "|:---|:---:|:---:|")
+    # Table contents
+    for op in operators
+        println(io, "| `$op` | ", 
+            method_overloaded(op, (Any, btype, Any, Any), ext), 
```

Comment:
I think this might be missing some overloads because the typing is more specific than `Any`

Reply:
I think it's due to `op!` having a different argument count than `op`.

Reply:
No, it's also because the method overloads won't accept Any in most places, for instance the extras have specific types, sometimes x and y are restricted to numbers or arrays, etc

Reply:
Basically you wanna check if for operator op, there is a method whose signature contains a backend argument more specific than the union of the backend type and its sparse counterpart

Reply:
The types are forwarded to `methods(f, (types), extension_module)`, which I think does the opposite:

```Julia
julia> methods(atan, (Float32, Float32))
# 1 method for generic function "atan" from Base:
 [1] atan(y::T, x::T) where T<:Union{Float32, Float64}
     @ Base.Math special/trig.jl:566

julia> methods(atan, (Any, Any))
# 8 methods for generic function "atan" from Base:
 [1] atan(a::Float16, b::Float16)
     @ Base.Math math.jl:1568
 [2] atan(y::BigFloat, x::BigFloat)
     @ Base.MPFR mpfr.jl:812
 [3] atan(::Missing, ::Missing)
     @ Base.Math math.jl:1584
 [4] atan(::Missing, ::Number)
     @ Base.Math math.jl:1586
 [5] atan(y::T, x::T) where T<:Union{Float32, Float64} # <--- (Float32, Float32)
     @ Base.Math special/trig.jl:566
 [6] atan(y::T, x::T) where T<:AbstractFloat
     @ Base.Math math.jl:875
 [7] atan(y::Real, x::Real)
     @ Base.Math math.jl:874
 [8] atan(::Number, ::Missing)
     @ Base.Math math.jl:1585
```

Reply:
Maybe try a top down approach, take all methods and filter by backend only

Reply:
This is what I'm already doing:
```Julia
ms = methods(f, argtypes, ext)

n = length(ms)
n == 0 && return "❌"
n == 1 && return "[✅]($(Base.url(only(ms))))"
return "[✅]($(Base.url(first(ms))))" # Optional TODO: return all URLs?
```